doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -522,28 +505,26 @@ the conversion in hypotheses :n:`{+ @ident}`.
    information about the constants it encounters and the unfolding decisions it
    makes.
 
-.. tacn:: red
-   :name: red
+.. tacn:: red {? @occurrences }
 
-   This tactic applies to a goal that has the form::
+   :math:`\beta`:math:`\iota`:math:`\zeta`-reduces `T`, if possible, in the selected
+   hypotheses and/or the goal, which must have the form:
 
-     forall (x:T1) ... (xk:Tk), T
+     :n:`forall @open_binders, T`
 
-   with :g:`T` :math:`\beta`:math:`\iota`:math:`\zeta`-reducing to :g:`c t`:sub:`1` :g:`... t`:sub:`n` and :g:`c` a
+   to :g:`c t`:sub:`1` :g:`… t`:sub:`n` where :g:`c` is a
    constant. If :g:`c` is transparent then it replaces :g:`c` with its
-   definition (say :g:`t`) and then reduces
-   :g:`(t t`:sub:`1` :g:`... t`:sub:`n` :g:`)` according to :math:`\beta`:math:`\iota`:math:`\zeta`-reduction rules.
+   definition and reduces again until no further reduction is possible.
 
-.. exn:: Not reducible.
-   :undocumented:
+   .. exn:: Not reducible.
+      :undocumented:
```

Comment:
I have no idea.  @Zimmi48?

Reply:
I don't know how to trigger it. It seems to be raised in cases similar to "No head constant to reduce" but only for tactic `fold` if we believe the code:

https://github.com/coq/coq/blob/16765871394a81975047b37f15a902fcc112dc40/pretyping/tacred.ml#L1208

And even for `fold`, I hit the other one:

```coq
Eval fold 0 in 0.
(* No head constant to reduce *)
```

Reply:
Hah, we have
https://github.com/coq/coq/blob/16765871394a81975047b37f15a902fcc112dc40/pretyping/tacred.ml#L1207-L1208
but 
https://github.com/coq/coq/blob/16765871394a81975047b37f15a902fcc112dc40/pretyping/tacred.ml#L960-L962
and hence the `user_err` here is dead code.

This dead code was already present when the functions were first defined in ae47a499e6dbf4232a03ed23410e81a4debd15d1.  However, at this point in time, the error messages were the same.  They got out of sync in d223f85d0fd57ce74dcdcc8690a36f1ef87b408d, when the error message in the live codepath was changed without changing the error message in the dead codepath.  @herbelin , since all of these commits are yours, can you comment on the intent of error messages here?  (Should we restore this error message, or remove the dead code path and remove the documentation of this message?)